File: examples/scripts/hash.ts
// Demonstrates crypto.hash.* — nine algorithms, all returning lowercase hex.
// Inputs are interpreted as UTF-8 byte sequences.

runtime.log("=== crypto.hash.* ===");
runtime.log("md5:     ", crypto.hash.md5("abc"));
runtime.log("sha1:    ", crypto.hash.sha1("abc"));
runtime.log("sha256:  ", crypto.hash.sha256("abc"));
runtime.log("sha384:  ", crypto.hash.sha384("abc"));
runtime.log("sha512:  ", crypto.hash.sha512("abc"));
runtime.log("sha3_256:", crypto.hash.sha3_256("abc"));
runtime.log("sha3_512:", crypto.hash.sha3_512("abc"));
runtime.log("blake3:  ", crypto.hash.blake3("abc"));
runtime.log("crc32:   ", crypto.hash.crc32("abc"));

// Known-vector sanity check.
runtime.assert.equal(
  crypto.hash.sha256("abc"),
  "ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad",
);

File: cmd/sercon/hash.go
package main

import (
	"crypto/md5"
	"crypto/sha1"
	"crypto/sha256"
	"crypto/sha512"
	"encoding/hex"
	"fmt"
	"hash/crc32"

	"github.com/dop251/goja"
	"golang.org/x/crypto/sha3"
	"lukechampine.com/blake3"
)

// hashNamespace builds the `crypto.hash.*` member map. Inputs are interpreted as
// UTF-8 byte sequences; outputs are lowercase hex strings. crc32 uses the
// IEEE polynomial (the universal default).
func hashNamespace(vm *goja.Runtime) map[string]any {
	stringInput := func(call goja.FunctionCall) []byte {
		arg := call.Argument(0)
		if arg == nil || goja.IsUndefined(arg) || goja.IsNull(arg) {
			panic(vm.NewTypeError("hash: argument 1 must be a string"))
		}
		return []byte(arg.String())
	}

	return map[string]any{
		"md5": func(call goja.FunctionCall) goja.Value {
			sum := md5.Sum(stringInput(call))
			return vm.ToValue(hex.EncodeToString(sum[:]))
		},
		"sha1": func(call goja.FunctionCall) goja.Value {
			sum := sha1.Sum(stringInput(call))
			return vm.ToValue(hex.EncodeToString(sum[:]))
		},
		"sha256": func(call goja.FunctionCall) goja.Value {
			sum := sha256.Sum256(stringInput(call))
			return vm.ToValue(hex.EncodeToString(sum[:]))
		},
		"sha384": func(call goja.FunctionCall) goja.Value {
			sum := sha512.Sum384(stringInput(call))
			return vm.ToValue(hex.EncodeToString(sum[:]))
		},
		"sha512": func(call goja.FunctionCall) goja.Value {
			sum := sha512.Sum512(stringInput(call))
			return vm.ToValue(hex.EncodeToString(sum[:]))
		},
		"sha3_256": func(call goja.FunctionCall) goja.Value {
			sum := sha3.Sum256(stringInput(call))
			return vm.ToValue(hex.EncodeToString(sum[:]))
		},
		"sha3_512": func(call goja.FunctionCall) goja.Value {
			sum := sha3.Sum512(stringInput(call))
			return vm.ToValue(hex.EncodeToString(sum[:]))
		},
		"blake3": func(call goja.FunctionCall) goja.Value {
			sum := blake3.Sum256(stringInput(call))
			return vm.ToValue(hex.EncodeToString(sum[:]))
		},
		"crc32": func(call goja.FunctionCall) goja.Value {
			sum := crc32.ChecksumIEEE(stringInput(call))
			return vm.ToValue(fmt.Sprintf("%08x", sum))
		},
	}
}
